proposals/function-pointers.md

```diff
@@ -1,60 +1,61 @@
 
 ## Motivation
-The motivations and background for this feature are described in the following issue (as is a 
-potential implementation of the feature): 
+
+The motivations and background for this feature are described in the following issue (as is a
+potential implementation of the feature):
 
 https://github.com/dotnet/csharplang/issues/191
 
 This is an alternate design proposal to [compiler intrinsics]
 (https://github.com/dotnet/csharplang/blob/master/proposals/intrinsics.md)
 
-## Detailed Design 
+## Detailed Design
 
 ### Function pointers
-The language will allow for the declaration of function pointers using the `func*` syntax. The full syntax is described
-in detail in the next section but it is meant to resemble the syntax used by `delegate` declarations.
+
+The language will allow for the declaration of function pointers using the `delegate*` syntax. The full syntax is described
+in detail in the next section but it is meant to resemble the syntax used by `Func` and `Action` type declarations.
 
 ``` csharp
 unsafe class Example {
-    delegate void DAction(int a);
-
-    void Example(DAction d, func* void(int) f) {
-        d(42);
+    void Example(Action<int> a, delegate*<int, void> f) {
+        a(42);
         f(42);
     }
 }
 ```
 
 These types are represented using the function pointer type as outlined in ECMA-335. This means invocation
-of a `func*` will use `calli` where invocation of a `delegate` will use `callvirt` on the `Invoke` method.
+of a `delegate*` will use `calli` where invocation of a `delegate` will use `callvirt` on the `Invoke` method.
 Syntactically though invocation is identical for both constructs.
 
 The ECMA-335 definition of method pointers includes the calling convention as part of the type signature (section 7.1).
-The default calling convention will be `managed`. Alternate forms can be specified by adding the appropriate modifier 
-after the `func*` syntax: `cdecl`, `fastcall`, `stdcall`, `thiscall` or `winapi`. Example:
+The default calling convention will be `managed`. Alternate forms can be specified by adding the appropriate modifier
+after the `delegate*` syntax: `managed`, `cdecl`, `stdcall`, `thiscall`, or `unmanaged`. Example:
 
 ``` csharp
 // This method will be invoked using the cdecl calling convention
-func* cdecl int(int value);
+delegate* cdecl<int, int>;
 
 // This method will be invoked using the stdcall calling convention
-func* stdcall int(int value);
+delegate* stdcall<int, int>;
 ```
 
-Conversions between `func*` types is done based on their signature including the calling convention. 
+Conversions between `delegate*` types is done based on their signature including the calling convention.
 
 ``` csharp
 unsafe class Example {
     void Conversions() {
-        func* int(int, int) p1 = ...;
-        func* managed int(int, int) p2 = ...;
-        func* cdecl int(int, int) p3 = ...;
+        delegate*<int, int, int> p1 = ...;
+        delegate* managed<int, int, int> p2 = ...;
```

Comment:
In my opinion, the calling convention appearing before the `delegate*` makes more sense, both gramatically and visually, as it's saying it's a `managed delegate pointer` rather than a `delegate pointer managed`

Reply:
I think the reason for this is mostly easy disambiguation. See https://github.com/dotnet/csharplang/issues/2917 for discussion.

Reply:
The latter also makes sense if you read it as `delegate pointer [of type] managed`

Reply:
My main issue is the type is usually the last thing that appears in a declaration, and in my opinion `delegate* managed<int>` implies you can just have a `managed<int>`, which you can't. The calling convention can be thought of as an attribute - attributes don't come after the type itself :P 

Same goes if you think about it like the `in`, `out`, and `ref` modifiers. It's unusual for the language to put the modifier after the type specifier, given that everywhere else the modifier is specified before the type.

Reply:
OTOH, for existing uses of `delegate` today `delegate` is the keyword, followed by the delegate definition. This is the same, where you have `delegate*` followed by the definition. We don't think that this is inconsistent with C# today.

Reply:
@333fred except that it's for an expression or declaration, the syntax for pointers is the whole type followed by an asterisk token.